Support rpm and deb packaging

syuu1228 opened the pull request:
Add build_rpm.sh and build_deb.sh build scripts to produce rpm & deb package.

closes scylladb/scylla-pkg#3714

closes #121

piodul commented:
Just to be clear, the failing clippy check has nothing to do with this change - we are using the newest version of the `clippy` linter and apparently new lints were introduced which our code does not satisfy.

It would probably be nice to fix them first to have green CI, but I personally don't think it's required.

Opened an issue about it: https://github.com/scylladb/cpp-rust-driver/issues/123

wprzytula commented on dist/debian/build_deb.sh:
```diff
@@ -0,0 +1,139 @@
+    esac
+done
+
+is_redhat_variant() {
+    [ -f /etc/redhat-release ]
+}
```
Can Debian be a Redhat variant? (isn't this a copypaste leftover?)

syuu1228 replied:
Since we use pbuilder which build package in ubuntu/debian chroot environment, we actually can build this in Fedora (may able to build our CI servers w/o using container).

wprzytula commented:
I've successfully packaged the driver into RPM and installed it on my Fedora 39. Congratulations, @syuu1228! :rocket:

Lorak-mmk commented on dist/debian/debian/rules:
```diff
@@ -0,0 +1,70 @@
+rustc = "$(CURDIR)/debian/.cargo/bin/rustc"
+rustdoc = "$(CURDIR)/debian/.cargo/bin/rustdoc"
+
+[profile.packaging]
+inherits = "release"
+opt-level = 3
+codegen-units = 1
+debug = 2
+strip = "none"
```
Imo you can add this profile to our `Cargo.toml` - it will be easier for developers to build and test using this profile.

syuu1228 replied:
Why I not added it into Cargo.toml was, 
- generating .cargo/config on-the-fly is the official way to build Rust based package on both rpm/deb so I followed it
- rpmbuild wants to extract its default parameters of CFLAGS/CXXFLAGS/LDFLAGS
But it seems our package actually does not use these flags on building code, so I guess we can just drop CFLAGS/CXXFLAGS/LDFLAGS and hardcode profile.packaging to Cargo.toml.
Fixed to do so.

Lorak-mmk commented on dist/redhat/scylla-cpp-rust-driver.spec:
```diff
@@ -0,0 +1,97 @@
+rustc = "$(pwd)/scylla-rust-wrapper/.cargo/bin/rustc"
+rustdoc = "$(pwd)/scylla-rust-wrapper/.cargo/bin/rustdoc"
+
+[profile.packaging]
+inherits = "release"
+opt-level = 3
+codegen-units = 1
+debug = 2
+strip = "none"
```
ditto

syuu1228 replied:
Fixed, same as above.

Lorak-mmk commented on dist/redhat/scylla-cpp-rust-driver.spec:
```diff
@@ -0,0 +1,97 @@
+debug = 2
+strip = "none"
+
+[env]
+CFLAGS = "%{build_cflags}"
+CXXFLAGS = "%{build_cxxflags}"
+LDFLAGS = "%{build_ldflags}"
```
`[env]` sets variables `for any process run by Cargo`. Why do we need those variables there?

syuu1228 replied:
Dropped, since it seems not used, as described above.
Maybe it will use Rust/C/C++ mixed code?

Lorak-mmk commented on dist/redhat/scylla-cpp-rust-driver.spec:
```diff
@@ -0,0 +1,97 @@
+CXXFLAGS = "%{build_cxxflags}"
+LDFLAGS = "%{build_ldflags}"
+
+[install]
+root = "%{buildroot}%{_prefix}"
```
According to https://doc.rust-lang.org/beta/cargo/reference/config.html this sets location for apps installed using `cargo install`. I've looked trough this PR and I don't see any place where we use `cargo install`. Why is it required, what am I missing?

syuu1228 replied:
These lines are copied from referenced setting (rpm macro), but we don't use `cargo install` so dropped them.

roydahan commented:
@syuu1228 where are we standing with this one?
Can you please address/reply the comments?

syuu1228 commented on README.md:
````diff
@@ -318,6 +318,37 @@ Now, use `--gtest_filter` to run certain integration tests:
+
+To build rpm package, run follow command:
+```shell
+./dist/redhat/build_deb.sh --target mantic
````
fixed

syuu1228 commented on README.md:
````diff
@@ -318,6 +318,37 @@ Now, use `--gtest_filter` to run certain integration tests:
+To build rpm package, run follow command:
+```shell
+./dist/redhat/build_rpm.sh --target rocky-8-x86_64
+```
````
fixed

syuu1228 commented on dist/debian/changelog.template:
```diff
@@ -0,0 +1,5 @@
+
+  * Initial release.
+
+ -- Takuya ASADA <syuu@scylladb.com>  Mon, 24 Aug 2015 09:22:55 +0000
```
The file is copied from scylla main repo, and we actually ignoring this changelog feature (we never update changelog, update version tag instead), but fixed date anyway.

syuu1228 commented on SCYLLA-VERSION-GEN:
```diff
@@ -0,0 +1,113 @@
+		-o|--output-dir)
+			OUTPUT_DIR="$2"
+			shift
+			shift
```
Fixed.
The script is copied from scylla main repo to generate package version-release parameters, I do not know why it written in this way.

syuu1228 commented on SCYLLA-VERSION-GEN:
```diff
@@ -0,0 +1,113 @@
+# Default scylla version tags
+VERSION="$(sed -n -e 's/^version = \"\(.*\)\"$/\1/p' scylla-rust-wrapper/Cargo.toml)"
+
+if test -f version
```
Fixed.
I do not know why it written in this way, for same reason above.

syuu1228 commented on SCYLLA-VERSION-GEN:
```diff
@@ -0,0 +1,113 @@
+#!/bin/sh
+
```
Changed to `#!/bin/bash -e`, same as other scripts

syuu1228 commented on SCYLLA-VERSION-GEN:
```diff
@@ -0,0 +1,113 @@
+		# This ensures that the package manager will select your custom
+		# package over the standard release.
+		SCYLLA_BUILD=0
+		SCYLLA_RELEASE=$SCYLLA_BUILD.$DATE.$GIT_COMMIT
```
This is added at:
https://github.com/scylladb/scylladb/commit/31c8e6885bb90002a4f68fb076f2f2ae29f9f11f

It is for building custom pacakge build like `1.6.0-1.avi.20160121.<githash>` which can upgrade official release `1.6.0-0.20160120.<githash>`.

syuu1228 commented on SCYLLA-VERSION-GEN:
```diff
@@ -0,0 +1,113 @@
+	if [ "$GIT_COMMIT" = "$GIT_COMMIT_FILE" ]; then
+		exit 0
+	fi
+fi
```
It was added at :https://github.com/scylladb/scylladb/commit/c44f347886c105e9119baee13994a224a95cdbd9
The issue was:
https://github.com/scylladb/scylla-pkg/issues/826

When we build multiple packages from same repository we run SCYLLA-VERSION-GEN multiple times.
It normally generates same version tags for each time, but `$DATE` may increment when the build started just before 12:00AM and did not finished until 12:00AM.
So the patch tries to skip generating version when git hash unchanged.

syuu1228 commented on dist/debian/debian/copyright:
```diff
@@ -0,0 +1,8 @@
+Format: http://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
+Upstream-Name: Scylla DB
+Upstream-Contact: http://www.scylladb.com/
+Source: https://github.com/scylladb/scylla
```
fixed

syuu1228 commented on dist/redhat/build_rpm.sh:
```diff
@@ -0,0 +1,83 @@
+# mock generates files owned by root, fix this up
+fix_ownership() {
+    sudo chown "$(id -u):$(id -g)" -R "$@"
+}
```
I think it was because root owned file can cause permission error on CI server.
This script copied from old scylla package build script, and related commit was:
https://github.com/scylladb/scylladb/commit/b167647bf6d81b6785d532790d6cb8c9bbe904e9

syuu1228 commented on dist/redhat/scylla-cpp-rust-driver.spec:
```diff
@@ -0,0 +1,97 @@
+opt-level = 3
+codegen-units = 1
+debug = 2
+strip = "none"
```
I think profile.packaging has `lto = true` since it `inherits = "release"`, no?

wprzytula replied:
https://doc.rust-lang.org/cargo/reference/profiles.html#release

As I read here, **`lto` is false by default in `release` profile**, probably because it greatly increases the build time.
As the packagers, however, we don't care too much about the build time - right?

syuu1228 replied:
> As I read here, **`lto` is false by default in `release` profile**

But our [profile.release] has `lto = true`:
https://github.com/scylladb/cpp-rust-driver/blob/master/scylla-rust-wrapper/Cargo.toml#L45
It will inherits to profile.packaging, no?

syuu1228 commented on dist/debian/debian/control:
```diff
@@ -0,0 +1,30 @@
+Depends: libscylla-cpp-driver0 (= ${binary:Version}),
+         ${misc:Depends},
+         ${shlibs:Depends},
+Description: Development libraries for ScyllDB Cpp-Rust Driver
```
fixed

syuu1228 commented on dist/debian/debian/control:
```diff
@@ -0,0 +1,30 @@
+               curl,
+               clang
+
+Package: libscylla-cpp-driver0
```
On Debian/Ubuntu, library package will follow SONAME of the shared library.
If SONAME is libxxhash.so.0 package name will be libxxhash0, and if SONAME is libuuid.so.1 package name will be libuuid1.
I think this is for compatibility, to libXXX0 and libXXX1 can co-exist which avoid breaking binaries when library changed ABI.

Detail of the package naming on libraries is described on Debian Policy: https://www.debian.org/doc/debian-policy/ch-sharedlibs.html

But the problem is our library actually does not have SONAME.
cdylib in Rust cannot specify SONAME when we build on Cargo:
https://github.com/rust-lang/cargo/issues/5045
Seems like some people using other build tool like cmake and running rustc directly to specify SONAME.

So we don't really have tail number for this package but I used "0" from current version "0.0.1".

This rule is not a must thing, we can drop "0" and use libcpp-rust-driver anyway.
But I guess lack of SONAME may cause more problems, and we should have it if it possible.

Actually Debian packaging system does not want install SONAME-less binary in /usr/lib, I got warning messages when I build the package.
I added workaround to ignore the message, for now.

wprzytula replied:
With Cargo you indeed cannot set SONAME, but according to https://github.com/rust-lang/cargo/issues/5045 it is possible to be done with a flag passed currently to `rustc`. Could you try experimenting with `rustc` invocations to see if it's possible to build our _cdylib_ with SONAME set?

syuu1228 replied:
Added support SONAME and versioned file name on .so.
I used patchelf to append SONAME on built .so file, since it seems easiest way to implement this.

Referenced the Rust based project on Mozilla:
https://github.com/mozilla/application-services/pull/175/files


The alternative way to implement this would be librsvg approach.
It build static C library on Cargo, and then run C linker to generate a shared library with SONAME.
Something like this:
```
build rsvg/librsvg-2.so.2.58.91: c_LINKER rsvg/librsvg-2.so.2.58.91.p/meson-generated_..__rsvg_dummy.c.o | /usr/lib64/libcairo-gobject.so /usr/lib64/libcairo.so /usr/lib64/libfontconfig.so /usr/lib64/libfreetype.so /usr/lib64/libgdk_pixbuf-2.0.so /usr/lib64/libgio-2.0.so /usr/lib64/libglib-2.0.so /usr/lib64/libgmodule-2.0.so /usr/lib64/libgobject-2.0.so /usr/lib64/libharfbuzz.so /usr/lib64/libm.so /usr/lib64/libpango-1.0.so /usr/lib64/libpangocairo-1.0.so /usr/lib64/libpangoft2-1.0.so /usr/lib64/libpng16.so /usr/lib64/libxml2.so rsvg/librsvg-2.0.def rsvg/librsvg-2.0.ver rsvg/librsvg_2.a
 LINK_ARGS = -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,-soname,librsvg-2.so.2 -Wl,--whole-archive -Wl,--start-group rsvg/librsvg_2.a -Wl,--no-whole-archive -Wl,--gc-sections -Wl,-Bsymbolic-functions /usr/lib64/libcairo.so /usr/lib64/libcairo-gobject.so /usr/lib64/libgobject-2.0.so /usr/lib64/libglib-2.0.so /usr/lib64/libpng16.so /usr/lib64/libfreetype.so /usr/lib64/libgdk_pixbuf-2.0.so /usr/lib64/libgio-2.0.so /usr/lib64/libharfbuzz.so /usr/lib64/libxml2.so /usr/lib64/libpangocairo-1.0.so /usr/lib64/libpango-1.0.so /usr/lib64/libpangoft2-1.0.so /usr/lib64/libfontconfig.so -Wl,--export-dynamic /usr/lib64/libgmodule-2.0.so -pthread -lm -Wl,--end-group
```

It's bit complecated since requires another build system outside of Cargo, and also need to take care dependencies.
reference: https://gitlab.gnome.org/GNOME/librsvg

syuu1228 commented:
Fixed codes mentioned on comments, also bash coding style fixed (by following [ShellCheck](https://github.com/koalaman/shellcheck) rules)

syuu1228 commented:
Just realized that generating pkgconfig files on deb was not implemented yet, fixed.

roydahan commented:
We would really want to merge and release 0.1.0 by the end of next week.
So, whoever should review/re-review and/or address comments, please prioritze.

mykaul commented on dist/common/pkgconfig/scylla_cpp_driver.pc.in:
```diff
@@ -0,0 +1,11 @@
+Version: @version@
+Libs: -L${libdir} -lscylla_cpp_driver
+Cflags: -I${includedir}/scylladb
+URL: https://github.com/scylladb/cpp-rust-driver
```
Do we need to add a license line here as well?

wprzytula replied:
ping @syuu1228

syuu1228 replied:
Well, I don't find license directive on[ pkg-config manual](https://linux.die.net/man/1/pkg-config), what does it means?

Lorak-mmk reviewed (commented):
I've built the rpm package and compared it to our cpp-driver rpm package (https://github.com/scylladb/cpp-driver/releases/download/2.16.2b/scylla-cpp-driver-2.16.2b-1.el7.x86_64.rpm ). I have  few questions (sorry if those are obvious, I don't have much experience in packaging).
cpp-driver provides the following files:
```
/usr/lib64/libscylla-cpp-driver.so
/usr/lib64/libscylla-cpp-driver.so.2
/usr/lib64/libscylla-cpp-driver.so.2.16.2-b
/usr/share/doc/scylla-cpp-driver-2.16.2b
/usr/share/doc/scylla-cpp-driver-2.16.2b/LICENSE.txt
/usr/share/doc/scylla-cpp-driver-2.16.2b/README.md
```

while cpp-rust-driver provides:
```
/usr/lib/.build-id
/usr/lib/.build-id/8a
/usr/lib/.build-id/8a/ad7ba567e0fd997cb3aa7c1819083466bfa71e
/usr/lib64/libscylla_cpp_driver.so
/usr/share/doc/scylla-cpp-rust-driver
/usr/share/doc/scylla-cpp-rust-driver/LICENSE
/usr/share/doc/scylla-cpp-rust-driver/README.md
```

1. What is the `.build-id` folder?
2. Why do we only have .so without version number in the name? cpp-driver provides, as you can see, `.so.<major-version>` and `.so.<full-version>`, which are symlinks to main `.so`. Why is that?
3. cpp-rust-driver has dashes in .so name while cpp-driver has underscores. Is it going to be a problem for a user looking for a drop-in replacement?

One more thing: apart from the commit with packaging there are 2 more commits. First one claims to update Rust Driver to newest version, but updates it to a commit from August of last year. Unless those 2 commits are necessary (and I don't see why would they be) please stick to the main goal of this PR which is packaging and drop those 2 commits. We'll update Rust Driver in a separate PR.

wprzytula commented:
> One more thing: apart from the commit with packaging there are 2 more commits. First one claims to update Rust Driver to newest version, but updates it to a commit from August of last year. Unless those 2 commits are necessary (and I don't see why would they be) please stick to the main goal of this PR which is packaging and drop those 2 commits. We'll update Rust Driver in a separate PR.

@Lorak-mmk These two commits are both mine, and they are already merged to master; I have no idea why they are displayed here.

Lorak-mmk commented:
> > One more thing: apart from the commit with packaging there are 2 more commits. First one claims to update Rust Driver to newest version, but updates it to a commit from August of last year. Unless those 2 commits are necessary (and I don't see why would they be) please stick to the main goal of this PR which is packaging and drop those 2 commits. We'll update Rust Driver in a separate PR.
> 
> @Lorak-mmk These two commits are both mine, and they are already merged to master; I have no idea why they are displayed here.

Rebasing a PR on master will probably fix this

wprzytula commented:
> cpp-rust-driver has dashes in .so name while cpp-driver has underscores. Is it going to be a problem for a user looking for a drop-in replacement?

Good catch! That's probably a very important question.

wprzytula commented on dist/redhat/scylla-cpp-rust-driver.spec:
```diff
@@ -0,0 +1,78 @@
+
+%install
+rm -rf %{buildroot}
+install -Dpm0644 scylla-rust-wrapper/target/packaging/*.a -t %{buildroot}%{_libdir}
+# We need to avoid dereference symlink, so can't use install here
+cp -a scylla-rust-wrapper/target/packaging/{*.so.*,*.so} %{buildroot}%{_libdir}
```
Why do we need to avoid dereferencing symlink?

syuu1228 replied:
Otherwise, all libscylla_cpp_driver.so, libscylla_cpp_driver.so.0, libscylla_cpp_driver.so.0.0.1 becomes real file, not symlink.
It also caused Debian package build failure since it expected the shared library is verioned and has symlink like libfoo.X -> libfoo.X.Y.Z.
(it failed to find the symlink and cause the error)

syuu1228 commented:
> 1. What is the `.build-id` folder?

It is the standard place to store debuginfo on newer distribution.
Both RedHat variants and Debian variants use same place.
reference: https://fedoraproject.org/wiki/RolandMcGrath/BuildID

> 2\. Why do we only have .so without version number in the name? cpp-driver provides, as you can see, `.so.<major-version>` and `.so.<full-version>`, which are symlinks to main `.so`. Why is that?

I implemented workaround patch for this:
https://github.com/scylladb/cpp-rust-driver/compare/5036de631e6271d15d419a327210b05df2ce418d..f552e9c62cc0d45ed074414a3d3f8aa69b1aabbf

On C/C++ project, build system automatically generate these things, but cdylib in Cargo does not, so we need to take care manually.
Not just about so.X and so.X.Y symlinks, cdylib in Cargo also does not have SONAME support, so I added workaround for this too.
thread: https://github.com/scylladb/cpp-rust-driver/pull/122#discussion_r1600976090

> 3\. cpp-rust-driver has dashes in .so name while cpp-driver has underscores. Is it going to be a problem for a user looking for a drop-in replacement?

I think both filename and SONAME should be identical as latest cpp-driver (libscylla-cpp-driver.so.2 -> libscylla-cpp-driver.so.2.16.2).


To get maximum compatibility with cpp-driver, maybe we should try librsvg approach ([described above]( https://github.com/scylladb/cpp-rust-driver/pull/122#discussion_r1600976090)).
If we can use cpp-drvier's CMakeLists.txt to link a shared library and install it, I guess we will get very similar output than original one.
And probably it can be implement by replacing C++ targets with ExternalProject_Add() which invoke cargo build to generate staticlib.

syuu1228 commented:
> 3\. cpp-rust-driver has dashes in .so name while cpp-driver has underscores. Is it going to be a problem for a user looking for a drop-in replacement?

Fixed this issue on versioning.sh. New file name becomes like this:
```
$ ls -l /usr/lib64/libscylla*
-rw-r--r-- 1 root root 7746332 Mar 28 09:00 /usr/lib64/libscylla_cpp_driver.a
lrwxrwxrwx 1 root root      30 Mar 28 09:00 /usr/lib64/libscylla-cpp-driver.so -> libscylla-cpp-driver.so.2.16.1
lrwxrwxrwx 1 root root      30 Mar 28 09:00 /usr/lib64/libscylla-cpp-driver.so.2 -> libscylla-cpp-driver.so.2.16.1
-rwxr-xr-x 1 root root 3193400 Mar 28 09:00 /usr/lib64/libscylla-cpp-driver.so.2.16.1
```

Note that I tried to rename this in Cargo.toml, but I find out it can't since Cargo reject that.
It says:
```
$ cargo build
error: failed to parse manifest at `/home/syuu/cpp-rust-driver.3/scylla-rust-wrapper/Cargo.toml`

Caused by:
  library target names cannot contain hyphens: scylla-cpp-driver
```

wprzytula commented:
The CI is failing due to new problems being detected by a new toolchain. #124 fixes them, so rebasing on it (once it get merged) should make the CI pass.

wprzytula commented:
> The CI is failing due to new problems being detected by a new toolchain. #124 fixes them, so rebasing on it (once it gets merged) should make the CI pass.

Merged. Please rebase on master @syuu1228.

syuu1228 commented:
Rebased.

yaronkaikov commented:
@Lorak-mmk Please review again

Lorak-mmk commented:
> > 3. cpp-rust-driver has dashes in .so name while cpp-driver has underscores. Is it going to be a problem for a user looking for a drop-in replacement?
> 
> Fixed this issue on versioning.sh. New file name becomes like this:
> 
> ```
> $ ls -l /usr/lib64/libscylla*
> -rw-r--r-- 1 root root 7746332 Mar 28 09:00 /usr/lib64/libscylla_cpp_driver.a
> lrwxrwxrwx 1 root root      30 Mar 28 09:00 /usr/lib64/libscylla-cpp-driver.so -> libscylla-cpp-driver.so.2.16.1
> lrwxrwxrwx 1 root root      30 Mar 28 09:00 /usr/lib64/libscylla-cpp-driver.so.2 -> libscylla-cpp-driver.so.2.16.1
> -rwxr-xr-x 1 root root 3193400 Mar 28 09:00 /usr/lib64/libscylla-cpp-driver.so.2.16.1
> ```
> 
> Note that I tried to rename this in Cargo.toml, but I find out it can't since Cargo reject that. It says:
> 
> ```
> $ cargo build
> error: failed to parse manifest at `/home/syuu/cpp-rust-driver.3/scylla-rust-wrapper/Cargo.toml`
> 
> Caused by:
>   library target names cannot contain hyphens: scylla-cpp-driver
> ```

You said you fixed the issue but the snippet you posted still has dashes instead of underscores - why?
I see dashes in versioning.sh and in the RPM I've built. I also don't see the .so without the version.
List of the files in the rpm:
```
/usr/lib/.build-id
/usr/lib/.build-id/51
/usr/lib/.build-id/51/dfc674c62e991c55ee5ea8fec011a8fe8d9ac3
/usr/lib64/libscylla-cpp-driver.so.2
/usr/lib64/libscylla-cpp-driver.so.2.16.1
/usr/share/doc/scylla-cpp-rust-driver
/usr/share/doc/scylla-cpp-rust-driver/LICENSE
/usr/share/doc/scylla-cpp-rust-driver/README.md
``

syuu1228 commented:
> You said you fixed the issue but the snippet you posted still has dashes instead of underscores - why?

I think it's opposite, it should be replace underscores with dashes, since we want to rename libscylla_cpp_driver.so.X.Y.Z to libscylla-cpp-driver.so.X.Y.Z.
But I found that I did not renamed static library yet, so I implemented it and pushed new version.

The new version also contains follow changes:
 - We did not changed static library filename (libscylla_cp_driver.a). Now it renamed to same as cpp-driver (libscylla-cpp-driver_static.a)
 - pkg-config file are broken, since we renamed library name. Applied dashes style name (libscylla-cpp-driver)
 - Change include path to original one, for compatibility (/usr/include)
 - Add Conflicts with cpp-driver rpm/deb package
 - Rebased with master

syuu1228 commented:
I found that versioning.sh will causes `mv: 'target/release/libscylla_cpp_driver.a' and 'target/release/libscylla-cpp-driver_static.a' are the same file` error when we build twice on same profile.
This is not occur on build_rpm.sh and build_deb.sh since both are running in temporary container, but if we build the library manually or calling versioning.sh from CMake will be problematic.

To avoid the error, stop calling `mv` and use `cp --remove-destination` to remove the destination file first.
Note: `ln` is already safe since `-f` option does same behavior (remove the destination file first).

Lorak-mmk commented on README.md:
````diff
@@ -318,6 +318,37 @@ Now, use `--gtest_filter` to run certain integration tests:
+
+To build deb package, run the following command:
+```shell
+./dist/redhat/build_deb.sh --target mantic
````
`redhat` -> `debian`

Lorak-mmk commented:
> > You said you fixed the issue but the snippet you posted still has dashes instead of underscores - why?
> 
> I think it's opposite, it should be replace underscores with dashes, since we want to rename libscylla_cpp_driver.so.X.Y.Z to libscylla-cpp-driver.so.X.Y.Z. But I found that I did not renamed static library yet, so I implemented it and pushed new version.
> 

Oh, that's right. In my initial comment about it I posted correct file lists and then mistakenly wrote "cpp-rust-driver has dashes in .so name while cpp-driver has underscores" while it was the other way around.
After bulding the package I see those files inside:
```
/usr/lib/.build-id
/usr/lib/.build-id/61
/usr/lib/.build-id/61/48a6727fa1f90104938a95ad2845314910309a
/usr/lib64/libscylla-cpp-driver.so.2
/usr/lib64/libscylla-cpp-driver.so.2.16.1
/usr/share/doc/scylla-cpp-rust-driver
/usr/share/doc/scylla-cpp-rust-driver/LICENSE
/usr/share/doc/scylla-cpp-rust-driver/README.md
```

`/usr/lib64/libscylla-cpp-driver.so` is still missing compared to cpp-driver.

Another small issue: after building the package with `./dist/redhat/build_rpm.sh --target fedora-39-x86_64` a `version` file is created in main folder of repo.
I don't fully understand why, I thought that the build was done in some sort of container.
Anyways, it should be either removed or added to .gitignore.


>>  What is the .build-id folder?

> It is the standard place to store debuginfo on newer distribution.
> Both RedHat variants and Debian variants use same place.

I don't see this folder in .deb packages. Is this expected?
Oh wait, I actually see it in dbgsym deb, I assume this is expected.


Another question, hopefully last one: In dev packages (both rpm and deb) I don't see versioned SOs (`.so.2`, `.so.2.16.1`), only the normal one (`.so`). Is this expected?

syuu1228 commented:
Add [a workaround patch for Ubuntu 24.04 build](https://github.com/scylladb/cpp-rust-driver/compare/788f9f9baa3df65002b2bc52e8974b77edae5db6..ede74e83405f20668d614178c3f20a8fd0cd0554), since Ubuntu 24.04 version of patchelf breaks libscylla-cpp-driver.so for some reason (probably a bug on patchelf).
It will cause error when stripping debug symbol since the ELF header is corrupted.

The error message is like this:
```
        readelf -W --section-headers debian/libscylla-cpp-driver-dev/usr/lib/x86_64-linux-gnu/libscylla-cpp-driver_static.a | sed -n '/^ *\[[ 0-9]*]/s/ *\[[ 0-9]*\] *//p' | awk 'BEGIN {rv=1} /^NULL/ {next} $1 ~ /^.(text|data|(preinit|init|fini)_array$)/ {if ($5 !~ /^0+$/) rv=0} END { exit rv}'
f132a9c6e662bd19d130b1e3ee434af824abbfe0
readelf: Error: no .dynamic section in the dynamic segment
        install -m0755 -d debian/.debhelper/libscylla-cpp-driver0/dbgsym-root/usr/lib/debug/.build-id/f1
        objcopy --only-keep-debug --compress-debug-sections debian/libscylla-cpp-driver0/usr/lib/x86_64-linux-gnu/libscylla-cpp-driver.so.2.16.1 debian/.debhelper/libscylla-cpp-driver0/dbgsym-root/usr/lib/debug/.build-id/f1/32a9c6e662bd19d130b1e3ee434af824abbfe0.debug
objcopy: warning: debian/libscylla-cpp-driver0/usr/lib/x86_64-linux-gnu/libscylla-cpp-driver.so.2.16.1 has a corrupt string table index
objcopy: unable to modify 'debian/libscylla-cpp-driver0/usr/lib/x86_64-linux-gnu/libscylla-cpp-driver.so.2.16.1' due to errors
dh_strip: error: objcopy --only-keep-debug --compress-debug-sections debian/libscylla-cpp-driver0/usr/lib/x86_64-linux-gnu/libscylla-cpp-driver.so.2.16.1 debian/.debhelper/libscylla-cpp-driver0/dbgsym-root/usr/lib/debug/.build-id/f1/32a9c6e662bd19d130b1e3ee434af824abbfe0.debug returned exit code 1
dh_strip: error: Aborting due to earlier error
make: *** [debian/rules:15: binary] Error 25
dpkg-buildpackage: error: fakeroot debian/rules binary subprocess returned exit status 2
debuild: fatal error at line 1184:
dpkg-buildpackage -us -uc -ui failed
```

This can be avoid by using Ubuntu 23.10's patchelf, the patch do that in versioning.sh.
(This should be temporary solution but I not find other way for now)

syuu1228 commented:
> `/usr/lib64/libscylla-cpp-driver.so` is still missing compared to cpp-driver.

Sorry I forgot to describe what is this.
On both RedHat and Debian, library packages does not include `libXXX.so` file on main package, it will installed in -devel (or -dev in Debian) package.

Debian Policy Manual says:
_The development package should contain a symlink for the associated shared library without a version number. For example, the libgdbm-dev package should include a symlink from /usr/lib/libgdbm.so to libgdbm.so.3.0.0. This symlink is needed by the linker (ld) when compiling packages, as it will only look for libgdbm.so when compiling dynamically._
https://www.debian.org/doc/debian-policy/ch-sharedlibs.html#development-files

Seems like cpp-driver packaging libscylla-cpp-driver.so in main package, but unless the difference causes compatibility problem, I think we should keep current packaging since it looks correct.

syuu1228 commented:
> Another question, hopefully last one: In dev packages (both rpm and deb) I don't see versioned SOs (`.so.2`, `.so.2.16.1`), only the normal one (`.so`). Is this expected?

I think this is same topic as [I described above](https://github.com/scylladb/cpp-rust-driver/pull/122#issuecomment-2141150681), -dev package will contain libXXX.so, and main package will contain libXXX.so.1.2.3 and libXXX.so.1.

syuu1228 commented:
I just found that we actually can fix SONAME using RUSTFLAGS, even we build Rust in Cargo:
https://developers.redhat.com/articles/2022/09/05/how-create-c-binding-rust-library#fixing_the_soname
I think this is correct approach than paching ELF header, and it also should fix Ubuntu 24.04 patchelf issue, so [pushed new version to do that](https://github.com/scylladb/cpp-rust-driver/compare/1b96858aeaecb58ee50e7a0ebc4a1c9fd1f1d70d..bd1782eb3df939a1ed32302ab1b62a36ba9a9316).

syuu1228 commented:
> If so, please fix the last tiny issue I mentioned (`version` file, which should be .gitignore'd / removed after build / not created) and we can merge.

Pushed new version to add `version` to .gitignore.

Lorak-mmk commented on dist/redhat/scylla-cpp-rust-driver.spec:
```diff
@@ -0,0 +1,85 @@
+set -euo pipefail
+
+%build
+if [ -z $RUSTFLAGS ]; then
+    RUSTFLAGS="-Copt-level=3 -Cdebuginfo=2 -Ccodegen-units=1 -Cstrip=none -Cforce-frame-pointers=yes --cap-lints=warn"
+fi
```
Either I didn't notice it before or it wasn't here before. All those parameters (apart from `--cap-lints=warn` and `-Cforce-frame-pointers=yes`) are set in packaging profile in Cargo.toml, why set them again here?

Lorak-mmk commented on dist/debian/debian/rules:
```diff
@@ -0,0 +1,53 @@
+SCYLLA_VERSION := $(shell cat version | awk -F'-' '{print $1}' | sed 's/-/~/')
+
+VERSION_MAJOR := $(shell sed -n -e 's/^#define CASS_VERSION_MAJOR \(.*\)/\1/p' include/cassandra.h)
+RUSTFLAGS := -C debuginfo=2 --cap-lints warn -C linker=x86_64-linux-gnu-gcc -C link-arg=-Wl,-Bsymbolic-functions -C link-arg=-Wl,-z,relro -Clink-arg=-Wl,-soname=libscylla-cpp-driver.so.$(VERSION_MAJOR)
```
Same here, `-C debuginfo=2` probably shouldn't be set here

syuu1228 replied:
The reason why these parameters are applying duplicated parameters is same, this RUSTFLAGS is taken from Debian package build system.
After dropping duplicated parameters like above, it will be:
```RUSTFLAGS := --cap-lints warn -C linker=x86_64-linux-gnu-gcc -C link-arg=-Wl,-Bsymbolic-functions -C link-arg=-Wl,-z,relro -Clink-arg=-Wl,-soname=libscylla-cpp-driver.so.$(VERSION_MAJOR)```

syuu1228 commented:
@Lorak-mmk [Pushed new version](https://github.com/scylladb/cpp-rust-driver/compare/06ceaabb9d5af7e92f8949f63cc3adf759a17f27..58f140830b74193afba9c7960b4187232e05e133) to change RUSTFLAGS on both rpm & deb.

roydahan commented:
@syuu1228 can you please check why CI fails?

syuu1228 commented:
@roydahan This is a bug on .github/workflows, sent fix at #127

roydahan commented:
@wprzytula would you want the honor to merge?
Once we merge, we need to officially build and release 0.1.0.

wprzytula commented:
> @wprzytula would you want the honor to merge? Once we merge, we need to officially build and release 0.1.0.

Gladly!
